### PR TITLE
Disallow retreat to enemy capital

### LINKED ISSUE #189: Retreat to an enemy capital is forbidden
image

### PR DESCRIPTION
**Pull request by gereon77**
Fixes #189

### CODE REVIEW

**Comment by gereon77** [comment]
Well.. rethinking the topic let me assume that robb stark still is valid to retreat to the enemy capital by this modification. So the check has to go to getvalidretreatregions also which makes it a bit more complicated.. 

I will move the logic to combat game state...

Interesting scenario as losing against any other card than robb will destroy the units immediately but Robb might offer a valid retreat region and the starting region is blocked by lowest casualties rule...

File: agot-bg-game-server/src/common/ingame-game-state/action-game-state/resolve-march-order-game-state/combat-game-state/post-combat-game-state/resolve-retreat-game-state/ResolveRetreatGameState.ts [hunk 1]
```diff
@@ -73,7 +74,12 @@ export default class ResolveRetreatGameState extends GameState<
             // The loser is the attacker...
             if (overridenChooser == null) {
                 // ...and the chooser was not overriden:
-                // The units retreat automatically to the region they came from.
+                // The units retreat automatically to the region they came from unless they attacked from a enemy capital and haven't left a power token
+                if(this.attackingUnitsCannotRetreat(this.combat.attackingRegion, this.postCombat.loser)) {
```
Comment (Longwelwind) [inline agot-bg-game-server/src/common/ingame-game-state/action-game-state/resolve-march-order-game-state/combat-game-state/post-combat-game-state/resolve-retreat-game-state/ResolveRetreatGameState.ts]: You can simply check if the `attackingRegion` is controlled by an other house than `attacker` at the moment the retreat must be made. If it isn't then, they must all be killed.

Comment (gereon77) [reply agot-bg-game-server/src/common/ingame-game-state/action-game-state/resolve-march-order-game-state/combat-game-state/post-combat-game-state/resolve-retreat-game-state/ResolveRetreatGameState.ts]: I don't think that will help as `getController()` will return the attacker as his units are still present in `attackingRegion` at this point.

Comment (Longwelwind) [reply agot-bg-game-server/src/common/ingame-game-state/action-game-state/resolve-march-order-game-state/combat-game-state/post-combat-game-state/resolve-retreat-game-state/ResolveRetreatGameState.ts]: And checking `superControlPowerToken`?

On top of that, I think if player A moves from the capital of player B to attack player C without leaving a Power token and loses, the code at the moment won't correctly kill all units.

Comment (gereon77) [reply agot-bg-game-server/src/common/ingame-game-state/action-game-state/resolve-march-order-game-state/combat-game-state/post-combat-game-state/resolve-retreat-game-state/ResolveRetreatGameState.ts]: > 
> 
> And checking `superControlPowerToken`?
exactly that's done in attackingUnitsCannotRetreat (which will be renamed to attackingRegionBlockedForRetreat() to support Robb) but I decided to do the checks in method as this is more readable than a complex if chain.

> 
> On top of that, I think if player A moves from the capital of player B to attack player C without leaving a Power token and loses, the code at the moment won't correctly kill all units.

Why?

Comment (gereon77) [reply agot-bg-game-server/src/common/ingame-game-state/action-game-state/resolve-march-order-game-state/combat-game-state/post-combat-game-state/resolve-retreat-game-state/ResolveRetreatGameState.ts]: I fixed it that Robb Stark won't be able to retreat to the startingRegion. In addition I modified the unit-loss-warning to support all possible cases though ship loss will come with a separate PR but then I remove the logic from there.

**Comment by gereon77** [comment]
I tried to move the logic to `getValidRetreatRegions()` but then I have to handle the finalChooser there and it won't simplify the logic at all :(

**Comment by gereon77** [comment]
In fact I realized a problem by the current implementation and we need to move attacker or defender handling there. Right now `getValidRetreatRegions()` which only handles `defendingRegion` is called before the difference between attacker and defender is handled. In case attacker attacks but the defending area returns no retreat region the units will be killed and/or Robb can't be used correctly in case of attacking.

**Comment by gereon77** [comment]
So I followed your idea to move the logic to getValidRetreatRegions. It looks a bit easier now but game logs for what's going on are missing. I will add them in a second commit.

File: agot-bg-game-server/src/common/ingame-game-state/action-game-state/resolve-march-order-game-state/combat-game-state/post-combat-game-state/resolve-retreat-game-state/ResolveRetreatGameState.ts [hunk 2]
```diff
@@ -185,19 +186,60 @@ export default class ResolveRetreatGameState extends GameState<
...
+                // He can decide to retreat the units to a valid region adjacent to the combat
+                if (attackingRegionIsBlockedForRetreat) {
+                    // But if attacking region is blocked it must be filtered out
+                    possibleRetreatRegions = possibleRetreatRegions.filter(r => r != attackingRegion);
+                }
+            }
+        }
+        /* else {
+
+        }
```
Comment (Longwelwind) [inline agot-bg-game-server/src/common/ingame-game-state/action-game-state/resolve-march-order-game-state/combat-game-state/post-combat-game-state/resolve-retreat-game-state/ResolveRetreatGameState.ts]: The commented code can be removed, I think. The comment below suffice.

**Comment by Longwelwind** [comment]
Nice!

About the logs, try to keep them to minimal. I usually aim for one log per action of a player. In this case, I think we could print:

* One when the retreat location has been chosen
* One when the loser has chosen their casualties.

**Comment by gereon77** [comment]
Done.

**Pull request merged**